dhall/src/Dhall.hs

```diff
@@ -94,14 +94,14 @@ module Dhall
     , Inject
     , inject
     , genericToDhall
-    , RecordInputType(..)
-    , inputFieldWith
-    , inputField
-    , inputRecord
-    , UnionInputType(..)
-    , inputConstructorWith
-    , inputConstructor
-    , inputUnion
+    , RecordEncoder(..)
+    , encodeFieldWith
+    , encodeField
+    , recordEncoder
+    , UnionEncoder(..)
+    , encodeConstructorWith
+    , encodeConstructor
```

Comment:
I'm still not sure about the `encode{Field,Constructor}With` names. It's really a bit unfair that the `FromDhall` side got to use `field` and `constructor`!

Reply:
Would `constructorEncoder[With]` and `fieldEncoder[With]` be any better? At least it would keep the `encode` name free for better uses…

Reply:
We could also consider creating separate modules for encoding and decoding. Then we could have `field[With]` and `constructor[With]`, and `union` and `record` in both modules.

Reply:
To decrease the unfairness, how about names like `to{Record, Field, Singleton}With` for the `ToDhall` side and `from{Record, Field, Singleton}With` for `FromDhall`?

About module for encoding and decoding, it's an out-there suggestion but, the way `tomland` encodes and decodes `toml` is with a bi-directional codec, something like 
```haskell
data Codec a = 
    Codec { fromDhall  :: Expr Src Void -> Extractor Src Void a
            toDhall :: a -> Expr Src Void
            dhallType :: Expr Src Void
            }
```
It's pretty nice to keep the symmetry, and this way you always have to implement both encoders and decoders at the same time.

Reply:
Let's maybe leave the lack of symmetry between record and union decoders and encoders for future work.